source/base/index_set.cc

```diff
@@ -28,6 +28,257 @@
+{
+  return &accessor;
+}
+
+bool IndexSet::IntervalIterator::operator == (const IndexSet::IntervalIterator & other) const
+{
+  return accessor == other.accessor;
+}
+
+bool IndexSet::IntervalIterator::operator != (const IndexSet::IntervalIterator & other) const
+{
+  return !(*this == other);
+}
+
+bool IndexSet::IntervalIterator::operator < (const IndexSet::IntervalIterator & other) const
+{
+  return accessor < other.accessor;
+}
+
+int IndexSet::IntervalIterator::operator - (const IndexSet::IntervalIterator & other) const
+{
+  Assert (accessor.index_set == other.accessor.index_set, ExcInternalError());
+
+  return accessor.range_idx - other.accessor.range_idx;
+}
+
+
+///// ElementIterator
+
+bool IndexSet::ElementIterator::is_valid() const
+{
+  return range_idx < index_set.ranges.size() && idx<index_set.ranges[range_idx].end;
+}
+
+IndexSet::ElementIterator::ElementIterator(const IndexSet &idxset, index_type range_idx, index_type index)
+  : index_set(idxset), range_idx(range_idx), idx(index)
+{
+  Assert(range_idx < index_set.ranges.size(), ExcMessage(""));
+  Assert(idx >= index_set.ranges[range_idx].begin, ExcMessage(""));
+  Assert(idx <= index_set.ranges[range_idx].end, ExcMessage(""));
+}
+
+IndexSet::ElementIterator::ElementIterator(const IndexSet &idxset)
+  : index_set(idxset), range_idx(0), idx(0)
+{
+  if (index_set.ranges.size()>0)
+    {
+      range_idx = index_set.ranges.size()-1;
+      idx = index_set.ranges[range_idx].end;
+    }
+}
+
+const IndexSet::index_type &
+IndexSet::ElementIterator::operator* () const
+{
+  Assert(range_idx < index_set.ranges.size(), ExcMessage(""));
+  return idx;
+}
+
+bool IndexSet::ElementIterator::operator == (const IndexSet::ElementIterator &other) const
+{
+  Assert (index_set == other.index_set, ExcInternalError());
+  return range_idx == other.range_idx && idx==other.idx;
+}
+
+void IndexSet::ElementIterator::advance ()
+{
+  Assert(is_valid(), ExcMessage(""));
+  if (idx < index_set.ranges[range_idx].end)
+    ++idx;
+  if (idx == index_set.ranges[range_idx].end)
+    {
+      if (range_idx < index_set.ranges.size()-1)
+        {
+          ++range_idx;
+          idx = index_set.ranges[range_idx].begin;
+        }
+    }
+}
+
+
+IndexSet::ElementIterator &
+IndexSet::ElementIterator::operator++ ()
+{
+  advance();
+  return *this;
+}
+
+IndexSet::ElementIterator
+IndexSet::ElementIterator::operator++ (int)
+{
+  IndexSet::ElementIterator it = *this;
+  advance();
+  return it;
+}
+
+bool IndexSet::ElementIterator::operator != (const IndexSet::ElementIterator & other) const
+{
+  return !(*this == other);
+}
+
+bool IndexSet::ElementIterator::operator < (const IndexSet::ElementIterator & other) const
+{
+  Assert (index_set == other.index_set, ExcInternalError());
+  return range_idx < other.range_idx || (range_idx == other.range_idx && idx<other.idx);
+}
+
+int IndexSet::ElementIterator::operator - (const IndexSet::ElementIterator & other) const
+{
```

Comment:
You probably need the return type to be the signed type that corresponds to `types::global_dof_index`. You also need to test for overflow since the signed type can not store the whole range.